Ghidra/Processors/CP1600/data/languages/CP1600.slaspec

```diff
@@ -0,0 +1,651 @@
+	branch_external    = (4, 4)
+	branch_condition   = (0, 3)
+	external_condition = (0, 3)
+	
+	opcode6_9          = (6, 9)
+	opcode3_9          = (3, 9)
+	opcode2_9          = (2, 9)
+	opcode1_9          = (1, 9)
+	opcode0_9          = (0, 9)
+;
+
+define token jump_token (32)
+	unused_jump = (26, 31)
+	reg24_25    = (24, 25)
+	address_hi  = (18, 23)
+	jump_type   = (16, 17)
+	address_lo  = (0, 9)
+;
+
+define token double16 (32)
+	unused_lo = (17, 31)
+	value_lo  = (16, 23)
+	unused_hi = (8, 15)
+	value_hi  = (0, 7)
+;
+
+define token immediate16 (16)
+	imm16  = (0, 15)
+	addr16 = (0, 15)
+;
+
+define context contextreg
+	doublebyte = (0, 0)
+;
+
+attach variables [ reg0_1        ] [ R0 R1 R2 R3             ];
+attach variables [ reg0_2 reg3_5 ] [ R0 R1 R2 R3 R4 R5 R6 R7 ];
+attach variables [ reg24_25      ] [ R4 R5 R6 R7             ]; # R7 is not used.
+
+################################################################
+
+jmpdest16:     reloc is address_hi & address_lo [ reloc = (address_hi << 10) + address_lo; ] { export *:2 reloc;                         }
+branchdest16:  reloc is branch_sign=0 ; imm16   [ reloc = inst_start + 2 + imm16;          ] { export *:2 reloc;                         }
+branchdest16:  reloc is branch_sign=1 ; imm16   [ reloc = inst_start + 2 - imm16;          ] { export *:2 reloc;                         }
+splitimm16:    split is value_hi & value_lo     [ split = (value_hi << 8) + value_lo;      ] { local tmp:2 = split & 0xFFFF; export tmp; }
+
+################################################################
+
+macro resultFlags(value) {
+	Z = value == 0;
+	S = value s< 0; 
+}
+
+macro addition(first, second) {
+	local tmpC = carry(first, second);
+	local tmpO = scarry(first, second);
+	first = first + second;
+	C = tmpC;
+	O = tmpO;
+	resultFlags(first);
+}
+
+macro readHalfDoubleWord(reg, dest) {
+	if (reg != R6) goto <read>;
+	reg = reg - 1;
+	<read>
+	local __tmp__:2 = reg;
+	local __ptr__:2 = *__tmp__;
+	dest = (*:2 (__ptr__)) & 0x00FF;
+	if ((reg != R4) || (reg != R5)) goto <done>;
+	reg = reg + 1;
+	<done>
+}
+
+macro readDoubleWord(reg, dest) {
+	local __tmpl__:2 = 0;
+	readHalfDoubleWord(reg, __tmpl__);
+	local __tmph__:2 = 0;
+	readHalfDoubleWord(reg, __tmph__);
+	dest = (__tmph__ << 8) | __tmpl__;
+}
+
+macro comparison(first, second) {
+	local __val__ = first - second;
+	O = sborrow(first, second);
+	C = first < second; 
+	resultFlags(__val__);
+}
+
+macro subtraction(first, second) {
+	local __val__ = first - second;
+	O = sborrow(first, second);
+	C = first < second; 
+	resultFlags(__val__);
+	first = __val__;
+}
+
+################################################################
+
+define pcodeop TerminateCurrentInterrupt;
+define pcodeop SoftwareInterrupt;
+
+################################################################
+
+:ADD addr16, reg0_2 is opcode3_9=0x0058 & reg0_2 ; addr16 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local ptr:2 = addr16;
+	addition(reg0_2, *(*:2 ptr));
```

Comment:
I would double check the use of  `*(*:2 x)`   this seems like it should only be `*:2 ptr`

Reply:
This sprung out of a misunderstanding of what the decompiler window outputs for these instructions, as it would just show data labels without having a type qualifier attached to them so I had no idea on whether the dereferencing step already occurred or not.